flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java

```diff
@@ -0,0 +1,316 @@
+
+	@Test
+	public void testCalculateTotalBuffersSize() throws IOException {
+		int numberOfRemoteChannels = 2;
+		int numberOfLocalChannels = 0;
+
+		int numberOfBufferPerChannel = 2;
+		int numberOfBuffersPerGate = 8;
+
+		NettyShuffleEnvironment network = new NettyShuffleEnvironmentBuilder()
+			.setNetworkBuffersPerChannel(numberOfBufferPerChannel)
+			.setFloatingNetworkBuffersPerGate(numberOfBuffersPerGate)
+			.build();
+
+		Tuple3<SingleInputGate, List<RemoteInputChannel>, List<LocalInputChannel>> tuple1 = buildInputGate(
+			network,
+			numberOfRemoteChannels,
+			numberOfLocalChannels);
+
+		SingleInputGate inputGate1 = tuple1.f0;
+
+		SingleInputGate[] inputGates = new SingleInputGate[]{inputGate1};
+		FloatingBuffersUsageGauge floatingBuffersUsageGauge = new FloatingBuffersUsageGauge(inputGates);
+		ExclusiveBuffersUsageGauge exclusiveBuffersUsageGauge = new ExclusiveBuffersUsageGauge(inputGates);
+		CreditBasedInputBuffersUsageGauge inputBufferPoolUsageGauge = new CreditBasedInputBuffersUsageGauge(
+			floatingBuffersUsageGauge,
+			exclusiveBuffersUsageGauge,
+			inputGates);
+
+		try (CloseableRegistry closeableRegistry = new CloseableRegistry()) {
+
+			closeableRegistry.registerCloseable(network::close);
+			closeableRegistry.registerCloseable(inputGate1::close);
+
+			assertEquals(numberOfBuffersPerGate, floatingBuffersUsageGauge.calculateTotalBuffers(inputGate1));
+			assertEquals(numberOfRemoteChannels * numberOfBufferPerChannel, exclusiveBuffersUsageGauge.calculateTotalBuffers(inputGate1));
+			assertEquals(numberOfRemoteChannels * numberOfBufferPerChannel + numberOfBuffersPerGate, inputBufferPoolUsageGauge.calculateTotalBuffers(inputGate1));
+		}
+	}
+
+	@Test
+	public void testExclusiveBuffersUsage() throws IOException {
+		int numberOfRemoteChannelsGate1 = 2;
+		int numberOfLocalChannelsGate1 = 0;
+		int numberOfRemoteChannelsGate2 = 1;
+		int numberOfLocalChannelsGate2 = 1;
+
+		int numberOfRemoteChannelsTotal = numberOfRemoteChannelsGate1 + numberOfRemoteChannelsGate2;
+		int numberOfInputGates = 2;
+
+		int buffersPerChannel = 2;
+		int extraNetworkBuffersPerGate = 8;
+
+		CloseableRegistry closeableRegistry = new CloseableRegistry();
+
+		NettyShuffleEnvironment network = new NettyShuffleEnvironmentBuilder()
+			.setNetworkBuffersPerChannel(buffersPerChannel)
+			.setFloatingNetworkBuffersPerGate(extraNetworkBuffersPerGate)
+			.build();
+
+		Tuple3<SingleInputGate, List<RemoteInputChannel>, List<LocalInputChannel>> tuple1 = buildInputGate(
+			network,
+			numberOfRemoteChannelsGate1,
+			numberOfLocalChannelsGate1);
+		Tuple3<SingleInputGate, List<RemoteInputChannel>, List<LocalInputChannel>> tuple2 = buildInputGate(
+			network,
+			numberOfRemoteChannelsGate2,
+			numberOfLocalChannelsGate2);
+
+		SingleInputGate inputGate1 = tuple1.f0;
+		SingleInputGate inputGate2 = tuple2.f0;
+
+		RemoteInputChannel remoteInputChannel1 = tuple1.f1.get(0);
+		RemoteInputChannel remoteInputChannel2 = tuple1.f1.get(1);
```

Comment:
It might use `List<RemoteInputChannel> remoteInputChannels = tuple1.f1` instead, then we could loop the channel to avoid `drainAndValidate` twice below.